engines/hypno/hypno.cpp

```diff
@@ -0,0 +1,482 @@
+
+	} else if (!_levels[name].arcade.background.empty()) {
+		debugC(1, kHypnoDebugArcade, "Executing arcade level %s", name.c_str());
+		_prefixDir = _levels[name].arcade.prefix;
+		if (!_levels[name].arcade.intro.empty()) {
+			MVideo v(_levels[name].arcade.intro, Common::Point(0, 0), false, true, false);
+			runIntro(v);
+		}
+		changeScreenMode("arcade");
+		runArcade(_levels[name].arcade);
+	} else if (!_levels[name].puzzle.name.empty()) {
+		debugC(1, kHypnoDebugScene, "Executing puzzle level %s", name.c_str());
+		if (!_levels[name].arcade.intro.empty()) {
+			MVideo v(_levels[name].arcade.intro, Common::Point(0, 0), false, true, false);
+			runIntro(v);
+		}
+		runPuzzle(_levels[name].puzzle);
+	} else {
+		debugC(1, kHypnoDebugScene, "Executing scene level %s", name.c_str());
+		resetSceneState();
+		_prefixDir = _levels[name].scene.prefix;
+		if (!_levels[name].scene.intro.empty()) {
+			MVideo v(_levels[name].scene.intro, Common::Point(0, 0), false, true, false);
+			runIntro(v);
+		}
+		changeScreenMode("scene");
+		runScene(_levels[name].scene);
+	}
+}
+
+void HypnoEngine::runIntro(MVideo &video) {
+	Common::Event event;
+	stopSound();
+	disableCursor();
+	playVideo(video);
+
+	while (!shouldQuit() && video.decoder) {
+		while (g_system->getEventManager()->pollEvent(event)) {
+			// Events
+			switch (event.type) {
+			case Common::EVENT_KEYDOWN:
+				if (event.kbd.keycode == Common::KEYCODE_ESCAPE) {
+					skipVideo(video);
+				}
+				break;
+
+			default:
+				break;
+			}
+		}
+
+		if (video.decoder) {
+			if (video.decoder->endOfVideo()) {
+				skipVideo(video);
+			} else if (video.decoder->needsUpdate()) {
+				updateScreen(video);
+				drawScreen();
+			}
+		}
+
+		g_system->updateScreen();
+		g_system->delayMillis(10);
+	}
+}
+
+void HypnoEngine::runPuzzle(Puzzle puzzle) { error("Function \"%s\" not implemented", __FUNCTION__); }
+void HypnoEngine::showCredits() { error("Function \"%s\" not implemented", __FUNCTION__); }
+
+void HypnoEngine::loadImage(const Common::String &name, int x, int y, bool transparent) {
+	debugC(1, kHypnoDebugMedia, "%s(%s, %d, %d, %d)", __FUNCTION__, name.c_str(), x, y, transparent);
+	Graphics::Surface *surf = decodeFrame(name, 0);
+	drawImage(*surf, x, y, transparent);
+}
+
+void HypnoEngine::drawImage(Graphics::Surface &surf, int x, int y, bool transparent) {
+	if (transparent) {
+		_compositeSurface->transBlitFrom(surf, Common::Point(x, y), surf.getPixel(0, 0));
+	} else
+		_compositeSurface->blitFrom(surf, Common::Point(x, y));
+}
+
+Common::File *HypnoEngine::fixSmackerHeader(Common::File *file) {
+	Common::String magic;
+	magic += file->readByte();
+	magic += file->readByte();
+	magic += file->readByte();
+	magic += file->readByte();
+
+	if (magic == "HYP2") {
+		ByteArray *data = new ByteArray();
+		data->push_back('S');
+		data->push_back('M');
+		data->push_back('K');
+		data->push_back('2');
+		while (!file->eos()) {
+			data->push_back(file->readByte());
```

Comment:
This could be horribly slow. Why not get the file size and read it all at once? Also, if you know file size, you may allocate the required buffer right away instead of the dynamic allocation byte by byte.